### ISSUE TITLE
onTokenRefresh() trying to register default SW instead the one passed to useServiceWorker()

### ISSUE THREAD

**Issue by coluccini** [firebase/firebase-js-sdk#1381]
### [REQUIRED] Describe your environment

  * Operating System version: Mac 10.14.1
  * Browser version: Chrome 70
  * Firebase SDK version: ^5.4.0
  * Firebase Product: messaging

### [REQUIRED] Describe the problem
I'm using a custom service worker to handle push notification, so I'm passing the registration to `messaging.useServiceWorker()`. Everything works fine but when `onTokenRefresh()` is called it throws a `messaging/failed-serviceworker-registration` error because is trying to register the `firebase-messaging-sw.js` file, which it doesn't exist, instead the one passed `useServiceWorker()`. 

I'm not sure if this it's happening every time `onTokenRefresh()` is called (cause I don't know how to manually trigger a refresh 😔) but it usually happens after leaving the web open in the background for a while.

#### Relevant Code:
I'm not doing any weird stuff, just:
```javascript
navigator.serviceWorker
  .register(swUrl)
  .then((registration) => {
    const messaging = firebase.messaging();
    messaging.useServiceWorker(registration);
    // ...
```
and

```javascript
messaging.onTokenRefresh(() => {
    saveDeviceToken(); // function that calls messaging.getToken() and send it to the server
  });
```
I have just one point where I register a SW and one point where I set onTokenRefresh listener (and it should be always be happening after the SW registration). 
The weird thing is that before setting `onTokenRefresh()` I'm using some other `messaging` methods and none of them are trying to register the service worker. Any ideas why this might be happening?
Thanks!

**Comment by mmermerkaya** [comment]
That sounds really weird, because `onTokenRefresh` doesn't do anything with the service worker. In fact, it is currently [unimplemented](https://github.com/firebase/firebase-js-sdk/blob/7a41fe71fd3effd71b61548546e290015f58456b/packages/messaging/src/controllers/window-controller.ts#L49) and your callback function should never be called.

I don't see anything unusual in the code you pasted. Have you defined any other callback handlers, like `onMessage`?

**Comment by coluccini** [comment]
I'm only using this method from `messaging`:

```javascript
messaging.getToken()
messaging.requestPermission()
messaging.onTokenRefresh()
```

It could be that the error is Threw cause, for some reason, `messaging` is getting called before `navigator.serviceWorker.register()` is called/finished? I'm using Create React App and the default code order is:

```javascript
ReactDOM.render(
  <Provider store={store()}>
    <App />
  </Provider>,
  document.getElementById('root'),
);

serviceWorker.register();
```

And the `messaging.onTokenRefresh()` method is called inside `<App />` (although in the same code `messaging.getToken()` is called before and I never saw the error over there 🤔).

Additional note: it happens more often on firefox than on Chrome…

**Comment by coluccini** [comment]
@mmermerkaya: I was wrong about the `onTokenRefresh()` method. I'm now pretty sure that the issue is triggered when there's a connectivity problem while calling the `messaging.getToken()` method. 

In the following video you can see that, in the first load, the custom SW passed to `useServiceWorker()`(`/local-service-worker.js`) is correctly called. Then I block connections to the domain `googleapis.com` (used to call `googleapis.com/identitytoolkit/v3/relyingparty/getAccountInfo?key=`) and firebase tries to register the default SW (`/firebase-messaging-sw.js`) instead of the custom SW. Everything works again after I unblock the domain connections...

The video: [https://www.youtube.com/watch?v=1gqXvZinaMo](https://www.youtube.com/watch?v=1gqXvZinaMo)

**Comment by jitendra1607** [comment]
/node_modules/@firebase/messaging/dist/index.cjs.js
You will find index.esm.js, index.cjs.js.map and index.esm.js.map also at above location.

var DEFAULT_SW_PATH = '/service-worker.js';
var DEFAULT_SW_SCOPE = '/';

Do above two changes wherever you find DEFAULT_SW_PATH and DEFAULT_SW_SCOPE
then build again, everything will work fine.

**Comment by DellaBitta** [comment]
Hi @coluccini,

It's been a while since this issue was created. Checking in if this is still a problem for you and if jitendra1607's suggestion has any behavioral changes for you. Thanks!

**Comment by coluccini** [comment]
Hi! It has been soooo long I don't even remember about this, but I've made a quick test (the same on the last video I shared) and everything seems to be working fine know (I think I have never tried jitendra1607's suggestion)

**Comment by DellaBitta** [comment]
Good to hear! I'm going to close this issue for now, but if it ever happens again, please create a new ticket. Feel free to link to this one if it's ever relevant. Thanks!
